client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,92 @@
+import { Controller } from '@hotwired/stimulus';
+
+/**
+ * <button
+ *   type="submit"
+ *   class="button button-longrunning"
+ *   data-controller="w-loader"
+ *   data-w-loader-active-class="button-longrunning-active"
+ *   data-w-loader-active-value="{% trans 'Signing in…' %}"
+ *   data-w-loader-duration-seconds-value="40"
+ *   data-action="w-loader#activate"
+ * >
+ *  {% icon name="spinner" %}
+ *  <em data-w-loader-target="label">{% trans 'Sign in' %}</em>
+ * </button>
+ */
+export class LoaderController extends Controller {
+  static classes = ['active'];
+  static targets = ['label'];
+  static values = {
+    active: { type: String, default: 'Loading' },
+    durationSeconds: { type: Number, default: 30 },
+  };
+
+  activeClass: string;
+  activeValue: string;
+  durationSecondsValue: number;
+  labelTarget: HTMLElement;
+  hasLabelTarget: boolean;
+
+  activate() {
+    let originalText;
+    if (this.hasLabelTarget === true) {
+      originalText = this.labelTarget.innerText;
+    }
+    originalText = this.element.textContent;
```

Comment:
You can use a class instance property for things like this OR a one liner to make it easier to read, also, this logic seems off - it looks like it will always set the originalText to the element text context.

```suggestion
    this.originalText = this.hasLabelTarget ? this.labelTarget.innerText : this.element.textContent;
    this.cancelSpinner(this.originalText)
```

```suggestion
    const originalText = this.hasLabelTarget ? this.labelTarget.innerText : this.element.textContent;
    this.cancelSpinner(originalText)
```

Note: Using a ternary statement (with the `?` & `:`) does not always make it easier to read, it depends on the context. However, when you set up a variable only to maybe adjust it based on a single truthy check, a ternary statement often reads better.